torch/lib/TensorLib/Scalar.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include<stdint.h>
+#include <stdexcept>
+#include <string>
+#include "TensorLib/HalfConvert.h"
+
+#ifdef TENSORLIB_CUDA_ENABLED
+#include <cuda_runtime.h>
+#include <cuda_fp16.h>
+#endif
+
+#if defined(__GNUC__)
+#define TLIB_ALIGN(n) __attribute__((aligned(n)))
+#elif defined(_WIN32)
+#define TLIB_ALIGN(n) __declspec(align(n))
+#else
+#define TLIB_ALIGN(n)
+#endif
+
+
+
+namespace tlib {
+
+
+template<typename To, typename From> To convert(From f) {
+  return static_cast<To>(f);
+}
+
+typedef struct  TLIB_ALIGN(2) {
+  unsigned short x;
+#ifdef TENSORLIB_CUDA_ENABLED
+  operator half() { return half { x }; }
+#endif
+  operator double();
+} Half;
+
+template<> Half convert(double f);
+template<> double convert(Half f);
+template<> Half convert(int64_t f);
+template<> int64_t convert(Half f);
+
+inline Half::operator double() {
+  return convert<double,Half>(*this);
+}
+#ifdef TENSORLIB_CUDA_ENABLED
+template<> half convert(double d);
+#endif
+
+#define TLIB_SCALAR_TYPES(_) \
+_(uint8_t,Byte,i) \
+_(int8_t,Char,i) \
+_(double,Double,d) \
+_(float,Float,d) \
+_(int,Int,i) \
+_(int64_t,Long,i) \
+_(int16_t,Short,i) \
+_(Half,Half,d)
+
+class Scalar {
```

Comment:
We use the largest possible storage (int64_t for integral types, double for fp) as our "Scalar" type. So any function that accepts a scalar will use this.